resources/session01/echo_server.py

```diff
@@ -32,32 +40,42 @@ def server(log_buffer=sys.stderr):
             #       the client so we can report it below.  Replace the
             #       following line with your code. It is only here to prevent
             #       syntax errors
-            addr = ('bar', 'baz')
+
+            conn, addr = sock.accept()
+            
             try:
                 print('connection - {0}:{1}'.format(*addr), file=log_buffer)
-
                 # the inner loop will receive messages sent by the client in
                 # buffers.  When a complete message has been received, the
                 # loop will exit
                 while True:
+                    data = conn.recv(16)
                     # TODO: receive 16 bytes of data from the client. Store
                     #       the data you receive as 'data'.  Replace the
                     #       following line with your code.  It's only here as
                     #       a placeholder to prevent an error in string
                     #       formatting
-                    data = b''
-                    print('received "{0}"'.format(data.decode('utf8')))
+                    print('received "{0}"'.format(data), file = log_buffer)
+
+                    if data:
+                         print('sending back to client', file = log_buffer)
+                         conn.sendall(data)
+                    else:
+                        msg = 'no more data from {0}:{1}'.format(*addr)
+                        print(msg, log_buffer)
+                        break
```

Comment:
The only criticism of this is that you will need to receive a 0-length message from the client in order to stop processing.  You are depending on the client behaving well.